spanner/benchmark/ycsb.py

```diff
@@ -0,0 +1,258 @@
+    print '[%s], AverageLatency(us), %f' % (
+        operation_upper, numpy.average(latencies_ms[operation]) * 1000.0)
+    print '[%s], LatencyVariance(us), %f' % (
+        operation_upper, numpy.std(latencies_ms[operation]) * 1000.0)
+    print '[%s], MinLatency(us), %f' % (
+        operation_upper, min(latencies_ms[operation]) * 1000.0)
+    print '[%s], MaxLatency(us), %f' % (
+        operation_upper, max(latencies_ms[operation]) * 1000.0)
+    print '[%s], 95thPercentileLatency(us), %f' % (
+        operation_upper,
+        numpy.percentile(latencies_ms[operation], 95.0) * 1000.0)
+    print '[%s], 99thPercentileLatency(us), %f' % (
+        operation_upper,
+        numpy.percentile(latencies_ms[operation], 99.0) * 1000.0)
+    print '[%s], 99.9thPercentileLatency(us), %f' % (
+        operation_upper,
+        numpy.percentile(latencies_ms[operation], 99.9) * 1000.0)
+    print '[%s], Return=OK, %d' % (operation_upper, op_counts[operation])
+    latency_array = numpy.array(latencies_ms[operation])
+    for j in range(num_bucket):
+      print '[%s], %d, %d' % (
+          operation_upper, j,
+          ((j <= latency_array) & (latency_array < (j + 1))).sum())
+    print '[%s], >%d, %d' % (
+          operation_upper, num_bucket, (num_bucket <= latency_array).sum())
+
+
+class WorkloadThread(threading.Thread):
+  """A single thread running workload."""
+
+  def __init__(self, database, parameters, total_weight, weights, operations):
+    threading.Thread.__init__(self)
+    self._database = database
+    self._parameters = parameters
+    self._total_weight = total_weight
+    self._weights = weights
+    self._operations = operations
+    self._latencies_ms = {}
+    for operation in self._operations:
+      self._latencies_ms[operation] = []
+
+  def run(self):
+    """Run a single thread of the workload."""
+    i = 0
+    operation_count = int(self._parameters['operationcount'])
+    while i < operation_count:
+      i += 1
+      weight = random.uniform(0, self._total_weight)
+      for j in range(len(self._weights)):
+        if weight <= self._weights[j]:
+          DoOperation(self._database, self._parameters['table'],
+                      self._operations[j], self._latencies_ms)
+          break
+
+  def latencies_ms(self):
+    """Returns the latencies."""
+    return self._latencies_ms
+
+
+def run_workload(database, parameters):
+  """Runs workload against the database."""
+  total_weight = 0.0
+  weights = []
+  operations = []
+  latencies_ms = {}
+  for operation in OPERATIONS:
+    weight = float(parameters[operation])
+    if weight <= 0.0:
+      continue
+    total_weight += weight
+    op_code = operation.split('proportion')[0]
+    operations.append(op_code)
+    weights.append(total_weight)
+    latencies_ms[op_code] = []
+
+  threads = []
+  start = timeit.default_timer()
+  for i in range(int(parameters['num_worker'])):
+    thread = WorkloadThread(database, parameters, total_weight, weights,
```

Comment:
I'm thinking if this would be better using multiprocess.Process because of the GIL.  I'm not sure yet.  I think it might depend on typical usage.

Reply:
@tseaver How does the database handle concurrency?

Reply:
does multiprocess.Process would mean that each process gets a new copy of the client with its own session pool, channel etc?

Reply:
Differ to vkedia for this.

Reply:
I believe it could be shared, but I'm not sure of a lot of things relating to this particular situation, I was just throwing it out there.  It might not even be worthwhile to investigate depending on usage.

Reply:
@tseaver @lukesneeringer to comment on what is the correct way to do this.
Essentially we want to run the benchmark with some fixed number of concurrent operations happening. How do we do that?

Reply:
Basically, python code is only ran one thread at a time.

Reply:
@vkedia @haih-g So I talked with @lukesneeringer earlier this morning.  He said that don't worry about the multithreading or multiprocessing issue.

Reply:
> Basically, python code is only ran one thread at a time.

@chemelnucfin Not so:  individual threads may release the GIL, e.g., when blocking for network / disk IO (these workers will presumably be doing a lot of network requests).

Reply:
they can release the GIL, but it's still running one thread at a time...
I guess if the bottleneck is the database, then it should be fine.
Thanks!

Reply:
Python threads can run at the same time on different processors, as long as at most *one* is holding the GIL.  Long-running webapp processes, for instance, make heavy use of that fact -- they are written to release it when using their ORMs / persistence machinery.

It is even possible to do computationally-intensive work in parallel, assuming that the C extensions being used know to release the GIL after pulling needed data from Python objects.

Reply:
interesting...  I've always used mp for that...  Thanks for the information!